### PR TITLE
Generate DeepCopy functions in the same package where the type is defined

### PR DESCRIPTION
**Pull request by wojtek-t**

### CODE REVIEW

**Comment by wojtek-t** [comment]
@lavalamp - This is something that I started doing before my paternity leave.
However, there is one issue here. Basically, this introduces circular dependencies due to pkg/api dependency (currently, to register deep-copy function, you need to call api.Scheme.Register....()

What do you think is the best way to solve it?

My idea is to:
- generate only deep copy function in the same directory where the type is defined,
- register ALL function in a single place (probably in pkg/api/register_generated_deep_copies.go or sth)
WDYT?

**Comment by wojtek-t** [comment]
OK - this is now ready for review

**Comment by wojtek-t** [comment]
@lavalamp - maybe you can delegate to someone? @caesarxuchao or @krousey ?

**Comment by lavalamp** [comment]
I'm crushed under reviews right now and I'm totally sitting on everything
that's not urgent. As long as there's no 1.2 fire drills, I hope to start
going through my backlog next week and reviewing or delegating as
appropriate. But it will be a bit slow, as you know googlers have extra
homework this week and next and I haven't even started. Thanks for bearing
with me.

File: pkg/util/sets/deep_copy_generated.go [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+// This file was autogenerated by deepcopy-gen. Do not edit it manually!
+
+package sets
+
+import (
+	conversion "k8s.io/kubernetes/pkg/conversion"
+)
+
+func DeepCopy_sets_Empty(in Empty, out *Empty, c *conversion.Cloner) error {
```
Comment (lavalamp) [inline pkg/util/sets/deep_copy_generated.go]: If this gets out of sync with the rest of the package, won't the tool fail to run?

File: pkg/util/sets/deep_copy_generated.go [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// This file was autogenerated by deepcopy-gen. Do not edit it manually!
+
```
Comment (lavalamp) [inline pkg/util/sets/deep_copy_generated.go]: Maybe we can come up with a build rule that causes the generator's compilation phase to ignore these files?

**Comment by lavalamp** [comment]
> However, there is one issue here. Basically, this introduces circular dependencies due to pkg/api dependency (currently, to register deep-copy function, you need to call api.Scheme.Register....()

Maybe, add a directory e.g. pkg/apimachinery/typetools/deepcopy? We should split the deep copy stuff out of scheme, there's no reason for it to be part of the scheme.

**Comment by lavalamp** [comment]
So basically my requirement here is that running the deep copy generator should never fail due to the deep copy generated code not compiling. I can imagine making a build rule, or somehow doing a two-phase thing where first you delete all of the old generated files, then you can actually compile and build the new ones.

**Comment by wojtek-t** [comment]
Yes - we can easily remove all the files at the beginning and regenerate them.
I will try to add this today.

File: pkg/util/sets/deep_copy_generated.go (hunk 2 above)
Comment (wojtek-t) [inline pkg/util/sets/deep_copy_generated.go]: Done.

File: pkg/util/sets/deep_copy_generated.go (hunk 1 above)
Comment (wojtek-t) [inline pkg/util/sets/deep_copy_generated.go]: Fixed by adding a build rule for it.

**Comment by wojtek-t** [comment]
>
Maybe, add a directory e.g. pkg/apimachinery/typetools/deepcopy? We should split the deep copy stuff out of scheme, there's no reason for it to be part of the scheme.

Yeah, but that's a bigger change and I wouldn't like to do it in this PR.


@lavalamp - PTAL

**Comment by wojtek-t** [comment]
@lavalamp - I checked that now, even if the generated file doesn't compile, it works fine.

So PTAL

File: cmd/libs/go2idl/deepcopy-gen/generators/deepcopy.go [hunk 3]
```diff
@@ -55,48 +56,52 @@ func DefaultNameSystem() string {
...
-	for _, inputDir := range arguments.InputDirs {
-		packages = append(packages,
-			&generator.DefaultPackage{
-				PackageName: filepath.Base(inputDir),
-				PackagePath: inputDir,
-				HeaderText: append(boilerplate, []byte(
-					`
+	header := append([]byte(
+		`
+// +build !ignore_autogenerated
```
Comment (lavalamp) [inline cmd/libs/go2idl/deepcopy-gen/generators/deepcopy.go]: Awesome!

**Comment by lavalamp** [comment]
The build rule looks pretty slick, I'm glad you were able to make that work! LGTM

**Comment by wojtek-t** [comment]
@lavalamp - yeah,  it took my some time to understand how exactly that works, but it seems this is exactly what we need.

Thanks for review!

**Comment by wojtek-t** [comment]
I regenerated the files - this should fix the unit test failures.

No code changes - reapplying lgtm label.

**Comment by lavalamp** [comment]
Awesome, thanks! Can't wait for the next round taking care of the
conversion functions!

**Comment by wojtek-t** [comment]
@lavalamp - https://github.com/kubernetes/kubernetes/pull/23110 roughly finishes migration for deep-copy
I already started some work around conversion, but I'm doing it in the background, so it will take make probably at least few more days.

**Comment by lavalamp** [comment]
No rush at all, we've lived with it for many months at this point...

**Pull request merged**